File: ElixirWar/lib/War.ex
defmodule War do
  @moduledoc """
    Documentation for `War`.
  """

  @doc """
    Function stub for deal/1 is given below. Feel free to add
    as many additional helper functions as you want.

    The tests for the deal function can be found in test/war_test.exs.
    You can add your five test cases to this file.

    Run the tester by executing 'mix test' from the war directory
    (the one containing mix.exs)
  """

  def deal(shuf) do
    players = setDeck(shuf)
    p1 = elem(players, 0)
    p2 = elem(players, 1)
    playGame(p1, p2, [], false)
  end

  def playGame([], [], warchest, _) do
    # when both players lose all cards at the same time
    denormalizeDeck(warchest)
  end
  def playGame(p1, [], warchest, _) do
    # when player 1 wins
    denormalizeDeck(p1++Enum.sort(warchest, :desc))
  end
  def playGame([], p2, warchest, _) do
    # when player 2 wins
    denormalizeDeck(p2++Enum.sort(warchest, :desc))
  end
  def playGame([h1 | t1], [h2 | t2], warchest, false) do
    # regular non-war turn
    cond do
      h1 == h2 ->
        # call play game again, with param set to true to indicate a war
        playGame(t1, t2, warchest++[h1]++[h2], true)
      h1 > h2 ->
        # when player 1 wins the turn add all cards to their deck
        playGame(t1++(Enum.sort(warchest++[h1]++[h2], :desc)), t2, [], false)
      h1 < h2 ->
        # when player 2 wins the turn add all cards to their deck
        playGame(t1, t2++(Enum.sort(warchest++[h1]++[h2], :desc)), [], false)
    end
  end
  def playGame([h1 | t1], [h2 | t2], warchest, true) do
    # during war remove one card from each deck and add it to the warchest
    playGame(t1, t2, Enum.sort(warchest++[h1]++[h2], :desc), false)
  end

  # converts aces to 13 so it is the highest card, and shifts all other cards down by 1
  def normalizeDeck(deck) do
    f1 = fn(a) ->
      if a == 1 do
        13
      else
        a - 1
      end
    end
    Enum.map(deck, f1)
  end

  # converts aces from 13 to 1, and shifts all other cards up by 1
  def denormalizeDeck(deck) do
    f1 = fn(a) ->
      if a == 13 do
        1
      else
        a + 1
      end
    end
    Enum.map(deck, f1)
  end

  # deals cards out to two lists
  def setDeck(deck) do
    setDeck(normalizeDeck(deck), [], [])
  end
  def setDeck([], p1, p2) do
    {p1, p2}
  end
  def setDeck([h | t], p1, p2) do
    setDeck(tl(t), [h | p1], [hd(t) | p2])
  end

end

File: HaskellWar/src/War.hs
module War (main, deal) where
import Data.List
import Debug.Trace

main :: IO ()
main = do
    putStrLn "Hello, world!"

{--
Function stub(s) with type signatures for you to fill in are given below. 
Feel free to add as many additional helper functions as you want. 

The tests for these functions can be found in src/TestSuite.hs. 
You are encouraged to add your own tests in addition to those provided.

Run the tester by executing 'cabal test' from the war directory 
(the one containing war.cabal)
--}
    
deal :: [Int] -> [Int]
deal shuf = do
    let p1 = getPlayer1(normalizeDeck shuf)
    let p2 = getPlayer2(normalizeDeck shuf)
    playGame p1 p2 [] False


playGame :: [Int] -> [Int] -> [Int] -> Bool -> [Int]
playGame [] [] warchest _ =
    -- when both players lose all cards at the same time
    denormalizeDeck warchest
playGame p1 [] warchest _ =
    -- when player 1 wins
    denormalizeDeck(p1++reverse(sort warchest))
playGame [] p2 warchest _ =
    -- when player 2 wins
    denormalizeDeck(p2++reverse(sort warchest))
playGame (h1:t1) (h2:t2) warchest False =
    -- regular non-war turn
    if h1 == h2 then
        -- call play game again, with param set to true to indicate a war
        playGame t1 t2 (warchest++[h1]++[h2]) True
    else if h1 > h2 then
        -- when player 1 wins the turn add all cards to their deck
        playGame (t1++reverse(sort(warchest++[h1]++[h2]))) t2 [] False
    else
        -- when player 2 wins the turn add all cards to their deck
        playGame t1 (t2++reverse(sort(warchest++[h1]++[h2]))) [] False
playGame (h1:t1) (h2:t2) warchest True =
    -- during war remove one card from each deck and add it to the warchest
    playGame t1 t2 (reverse(sort(warchest++[h1]++[h2]))) False

-- converts all cards in deck; aces from 1 to 13 so it is the highest card, and shifts all other cards down by 1
normalizeDeck :: [Int] -> [Int]
normalizeDeck [] = []
normalizeDeck deck =
    map normalizeCard deck

-- takes a single card; aces to 13 so it is the highest card, and shifts all other cards down by 1
normalizeCard :: Int -> Int
normalizeCard card =
    if card == 1 then 13
    else card - 1

-- converts all cards in a deck; aces from 13 to 1, and shifts all other cards up by 1
denormalizeDeck :: [Int] -> [Int]
denormalizeDeck [] = []
denormalizeDeck deck =
    map denormalizeCard deck

-- takes a single card; aces from 13 to 1, and shifts all other cards up by 1
denormalizeCard :: Int -> Int
denormalizeCard card =
    if card == 13 then 1
    else card + 1

-- returns player1's deck, every other card starting from the first card
getPlayer1 :: [Int] -> [Int]
getPlayer1 [a] = [a]
getPlayer1 [a, _] = [a]
getPlayer1 (xh:xt) = getPlayer1 (tail xt) ++ [xh]

-- returns player2's deck, every other card starting from the second card
getPlayer2 :: [Int] -> [Int]
getPlayer2 [a] = []
getPlayer2 [_, a] = [a]
getPlayer2 (xh:xt) = getPlayer2 (tail xt) ++ [head xt]